tests/test_models.py

```diff
@@ -1,37 +1,252 @@
+        self.assertEqual(customers[0].id, original_id)
+        self.assertEqual(customers[0].email, "mya6511@nyu.edu")
+        
+    def test_update_no_id(self):
+        """It should not Update a Customer with no id"""
+        customer = CustomerFactory()
+        logging.debug(customer)
+        customer.id = None
+        self.assertRaises(DataValidationError, customer.update)
+    
+    def test_delete_a_customer(self):
+        """It should Delete a Customer"""
+        customer = CustomerFactory()
+        customer.create()
+        self.assertEqual(len(Customer.all()), 1)
+        # delete the customer and make sure it isn't in the database
+        customer.delete()
+        self.assertEqual(len(Customer.all()), 0)
+    
+    def test_list_all_customers(self):
+        """It should List all Customers in the database"""
+        customers = Customer.all()
+        self.assertEqual(customers, [])
+        # Create 5 Customers
+        for _ in range(5):
+            customer = CustomerFactory()
+            customer.create()
+        # See if we get back 5 Customers
+        customers = Customer.all()
+        self.assertEqual(len(customers), 5)
+    
+    def test_serialize_a_customer(self):
+        """It should serialize a Customer"""
+        customer = CustomerFactory()
+        data = customer.serialize()
+        self.assertNotEqual(data, None)
+        self.assertIn("id", data)
+        self.assertEqual(data["id"], customer.id)
+        self.assertIn("first_name", data)
+        self.assertEqual(data["first_name"], customer.first_name)
+        self.assertIn("last_name", data)
+        self.assertEqual(data["last_name"], customer.last_name)
+        self.assertIn("email", data)
+        self.assertEqual(data["email"], customer.email)
+        self.assertIn("password", data)
+        self.assertEqual(data["password"], customer.password)
+    
+    def test_deserialize_a_customer(self):
+        """It should de-serialize a Customer"""
+        data = CustomerFactory().serialize()
+        customer = Customer()
+        customer.deserialize(data)
+        self.assertNotEqual(customer, None)
+        self.assertEqual(customer.id, None)
+        self.assertEqual(customer.first_name, data["first_name"])
+        self.assertEqual(customer.last_name, data["last_name"])
+        self.assertEqual(customer.email, data["email"])
+        self.assertEqual(customer.password, data["password"])
+    
+    def test_deserialize_missing_data(self):
+        """It should not deserialize a Customer with missing data"""
+        data = {"id": 1, "first_name": "Kitty", "email": "kitty@gmail.com"}
+        customer = Customer()
+        self.assertRaises(DataValidationError, customer.deserialize, data)
+    
+    def test_deserialize_bad_data(self):
+        """It should not deserialize bad data"""
+        data = "this is not a dictionary"
+        customer = Customer()
+        self.assertRaises(DataValidationError, customer.deserialize, data)
+
```

Comment:
Can add tests for deserializing data with key error (empty JSON or JSON without an id) and data which is not JSON (type error).

Reply:
Covered by the tests already.